### ISSUE TITLE
route_allfastest not working as expected?

### ISSUE THREAD

**Issue by sevrinrobstad** [memcached/memcached-proxylibs#11]
Hi again.

We are approaching taking memcached-proxy to production, but some testing is giving us some unexpected results.

We route all commands except for GET/GETS to route_allfastest. (see full config below)
By my understanding, route_allfastest should "OK" a write-command as long as at least one backend responds.
But when I restart one of the memcache-nodes, I get a couple of "SERVER_ERROR backend failure" when I set a test key, probably until the proxy mark the backend as down. 

Shouldn't this work? 
running latest memcached-proxy tag v1.6.31

```
verbose(true)
debug(true)

settings{
    active_request_limit = 5000,
    backend_connect_timeout = 0.8,
    backend_failure_limit = 2,
}
pools{
    write1 = {
        backends =  {
            "dev-memcache1:11211",
        }
    },
    write2 = {
        backends =  {
            "dev-memcache2:11211",
        }
    },
    read = {
        backends = {
            "dev-memcache1:11211",
            "dev-memcache2:11211",
        }
    },
}

routes{
    conf = {
        mode = "prefix",
        stop = "-"
    },
-- default handling
    default = route_allfastest{
        children = { "write1", "write2" },
        stats = true,
    },
-- override get/gets
    cmap = {
         get = route_failover{
             children = { "read" },
             stats = true,
             miss = true,
             shuffle = true,
             stats = true,
         },
         gets = route_failover{
             children = { "read" },
             stats = true,
             miss = true,
             shuffle = true,
             stats = true,
         },
    },
}
```

**Comment by dormando** [comment]
I'll take a look at this today. I wrote the behavior of that route to match mcrouter's allfastest. I'll double check what they do and either fix that, add an option or new route that's the "first ok" or "first good" response, which should avoid the backend errors.

**Comment by dormando** [comment]
Fwiw, based on your config I'm guessing you intended to do something like this:
```
pools{
    set_all = {
        { backends =  {
            "dev-memcache1:11211",
        } },
        { backends =  {
            "dev-memcache2:11211",
        } }
    }
}

routes{
    conf = {
        mode = "prefix",
        stop = "-"
    },
-- default handling
    default = route_allfastest{
        children = "set_all",
        stats = true,
    },
-- override get/gets
    cmap = {
         get = route_failover{
             children = "set_all",
             stats = true,
             miss = true,
             shuffle = true,
             stats = true,
         },
         gets = route_failover{
             children = "set_all",
             stats = true,
             miss = true,
             shuffle = true,
             stats = true,
         },
    },
}
```

The failover route works _by pool_. Since you were giving it a single pool with two backends in it, it won't fail over to anything.

In this config we create a magic pool set: two pools with one backend each in a list. We then pass that to everything. Failover randomize the list of pools (though only once on startup) as you requested (shuffle = true), and will actually try the other pool if it gets a miss.

**Comment by dormando** [comment]
According to mcrouter docs allsync is supposed to return the "worst" reply: https://github.com/facebook/mcrouter/wiki/List-of-Route-Handles#allsyncroute

So I'll look into adding at least modes where you return best or mid, unless this makes sense to you.

**Comment by dormando** [comment]
Seems my allfastest implementation just returns the first result. so I should update this a bit more anyway

**Comment by dormando** [comment]
derr I read the wrong link: https://github.com/facebook/mcrouter/wiki/List-of-Route-Handles#allfastestroute first non-error. I can fix that quickly

**Comment by dormando** [comment]
@sevrinrobstad should be fixed now. Added some failing tests (yes, routelib has tests!) then made them pass.

**Comment by sevrinrobstad** [comment]
Thank you so much for the update.

We try to wrap our heads around the "magic pool" stuff, and hope we have concluded correctly by assuming 
that the "magic pool" works like two pools when set up like you do here with two separate backends blocks in a list, 

So 

Two poolnames, one backend each:
```
pools{
    write1 = {
        backends =  {
            "dev-memcache1:11211",
        }
    },
    write2 = {
        backends =  {
            "dev-memcache2:11211",
        }
    },
  
```  
  
Would work the same as 

One poolname, two backends in separate "backends="
```
pools{
    write1 = {
        backends =  {
            "dev-memcache1:11211",
        },
        backends =  {
            "dev-memcache2:11211",
        }
    },

Will test and report back :)

**Comment by dormando** [comment]
The magic pools thing requires the name begin with `set_`. it's just shorthand for organizing pools.

IE: if you define pools pool1 and pool2 separately, you need to specify both in `route_failover`:
`get = route_failover{ children = { "pool1", "pool2" } }`

If you use the set syntax, ie: `set_pools = { { first pool definition }, { second pool definition } }` then down in `route_failover`:
`get = route_failoer{ children = "set_pools" }` here routelib will magically translate "set_pools" into `children = { "etc", "etc" }`  except we didn't need to give the pools individual names.

In you original configuration you were attempting to tell the failover code to "fail over" between the two backends defined in one pool, but the proxy cannot do that, so your config wouldn't work. All routes handlers only take pools as arguments: so you need to give failover two pools to try. In the example I fixed this with two pools. Each pool has one backend host.

Then, using `set_poolname` syntax I gave all of your route handlers the same arguments. Then you don't need to define the backends twice and can use the same name for everything.

It might be more clear if I give the pools names:
```
pools{
    set_all = {
        pool1 = { backends =  {
            "dev-memcache1:11211",
        } },
        pool2 = { backends =  {
            "dev-memcache2:11211",
        } }
    }
}
```

This is nearly equivalent to the fixed example I gave you, except that the pools are in a hash map instead of an ordered array. Which is thankfully not an important detail here :) We normally use pool sets with names for the zoned forms of the route handlers (ie; near and far zones)

In your original config you have write and read pools with the same hosts inside of them, but I think that was just due to a misunderstanding of how the failover route works?

**Comment by sevrinrobstad** [comment]
Ah, i thought "set_all" was just your name for the magic pool, but it's a SET of pools :)
I get it, thank you - will testing and report back!

The write and read pools were defined with the same hosts just as an example for future use, for example to read only from one Availability Zone.

**Comment by sevrinrobstad** [comment]
Hmm - Is this a bug?
When the route_failover children is set to "set_all", 
trying to GET an unset key returns like this
```
get test5
VALUE test5 0 1
5
END
get test
SERVER_ERROR lua failure
```
but when set up like I did before looks like this:
```
get test6
VALUE test6 0 1
6
END
get test
END
```

**Comment by sevrinrobstad** [comment]
Full test-config to spot potential typos:

```
verbose(true)
debug(true)

settings{
    active_request_limit = 5000,
    backend_connect_timeout = 0.8,
    backend_failure_limit = 2,
}

pools{
    set_all = {
        { backends =  {
            "dev-memcache1:11211",
        } },
        { backends =  {
            "dev-memcache2:11211",
        } }
    },
}

routes{
    conf = {
        mode = "prefix",
        stop = "-"
    },
-- default handling - write to all, return first good result
    default = route_allfastest{
        children = "set_all",
        stats = true,
    },
-- override get/gets - read from first (shuffled at start), fail over to next pool until first good result, if no good return last.
    cmap = {
         get = route_failover{
             children = "set_all",
             stats = true,
             miss = true,
             shuffle = true,
         },
         gets = route_failover{
             children = "set_all",
             stats = true,
             miss = true,
             shuffle = true,
         },
    },
}
```

**Comment by sevrinrobstad** [comment]
Basic testing confirms that all_fastest is fixed, I can not trigger errors when rebooting one of the two test memcache nodes anymore. GREAT!

Tested the bug (?) using set_all in route_failover again, and still gets the "lua failure" error when trying to get an unset key, so I have reverted back to defining multiple pools again. If this gets any updates, I'll be happy to test again.

**Comment by dormando** [comment]
Shoot sorry, I forgot to test that today. I'll get back to you in the next day.

For what it's worth, to get the full output of a lua error you need to check the log stream (for security reasons), which is a little awkward. Make another connection and run `watch proxyevents` to turn on th stream, then attempt to make requests until it makes that lua error.

I'm guessing it's some bug on my side but for future reference if you get 'lua error' back the full info is in there.

**Comment by dormando** [comment]
Just pushed a fix for the failover route. This does have tests so I'll check how that slipped through shortly as well.

Thanks for your patience!

**Comment by sevrinrobstad** [comment]
Well, thanx for the quick fixes. I'll test and report back here through the day.
I'm also opening another issue in the main repo, We have a segfault..

**Comment by dormando** [comment]
Thanks. I didn’t get a chance to check the tests on account of a local heat wave, will check soon.

sorry about the segfault. that’s odd. I swear it’s been in heavy production use for years without much issue with stability :) The startup code can be fragile so hopefully it’s just that

**Comment by dormando** [comment]
Fwiw: All the failover tests were explicitly setting the `failover_count`. This was a bug in calculating the failover_count automatically so that answers that. I'll add some more tests there and check the rest of the routes.

**Comment by sevrinrobstad** [comment]
I'll close this then, the OG issue is fixed, and the extra bug with set_all should be OK (although I haven't tested yet)
